### ISSUE TITLE
Multiple Daphne Processes?

### ISSUE THREAD

**Issue by claytondaley** [zerok/channels-in-docker#5]
We've been running Django Channels using a similar layout as this repository.  In Channels 1.0, it was trivial to get a single container to run multiple workers because Channels 1.0 separated the request handling from the workers.

With the transition to Channels 2.0, this separation has gone away.  The Channels documentation has an example of [multiple workers using channels + nginx](https://channels.readthedocs.io/en/latest/deploying.html?highlight=nginx#nginx-supervisor-ubuntu), but this configuration seems to require communication over a unix socket.

Do you have any thoughts on refining your example to show how to support multiple Daphne processes?

 - Is this possible to use a unix socket when nginx is hosted in a sidecare container?
 - If not, you would need a copy of nginx on the Daphne container to pull this off.
 - Once you're running nginx on the Daphne container, it's pretty easy to serve static files as well.
 - An nginx sidecar could still make sense for SSL termination, but may be unnecessary -- especially in cases like AWS ECS since the load balancer can do SSL termination.

**Comment by claytondaley** [comment]
I finally got the example config working for me so I can answer my own questions:

 - the `fcgi-program` directive actually binds to the host port so a local `nginx` process is _not_ required
 - this means that you can use it with an `nginx` sidecar -- for SSL termination and/or static file delivery

I had issues because the folder they use for the sockets (i.e. `/run/daphne/`) did not exist on my system.  This caused the daphne processes to exit (and eventually fail) with a status code of 0.  Unfortunately, no messages (error or otherwise) were logged by the process (even in verbose mode) to explain why.

To get this working with Docker, it seems to be enough to:

 - use the `supervisor.conf` directive found in the documentation (along with a bunch of standard sections like `unix_http_server`, `supervisord`, `rpcinterface:supervisor` and `supervisorctl`)
 - create (or change) the folder used for the `sock` files
 - start the Docker container (with a `CMD` or `run`) that involves `/usr/bin/supervisord` and referencing the `supervisor.conf` above (custom `conf` files can be used with the flag `--configuration=my_supervisor.conf`)
 - make sure supervisor doesn't daemonize by either:
    - adding `nodaemon=true` to the `[supervisord]` section of `supervisor.conf` (or)
    - using the `--nodaemon` flag when starting the Docker container (e.g. in the CMD)

The docker container will exit if the program its running returns.  If supervisor daemonizes, docker gets confused, thinks the process is done, and shuts the container down.

**Comment by claytondaley** [comment]
I have to backtrack some because it looked like it was working until I tried to actually use it.  I'm not able to connect to daphne from another container (and am using an on-container curl) so a local copy of nginx may be required.  My speculation is that sockets are only accessible locally (even if they appear to be bound to a "port") but need to do more work/testing to verify.

**Comment by claytondaley** [comment]
OK.  The issue is that I followed the instructions to bind the socket to `localhost:80`.  If you instead use `0.0.0.0:80`, it is available to remote callers (including a sidecar).  Using that configuration, a local copy of `nginx` is not necessary.

So my original question is still valid... any interest in updating this example to support multiple processes.  It should be possible without a local version of `nginx` provided you use `0.0.0.0:80`.

**Comment by zerok** [comment]
Hi :) Sorry for the late response! Yes, I will look into extending the example to also incorporate multiple daphne instances. Personally, though, I'm not a big fan of using something like supervisord for proxy-setups. If time permits I will look into getting this setup to work with either haproxy or envoy inside docker-compose in addition to the supervisord setup suggested in the documentation.

Sadly, I've quite a lot on my TODO list right now but I will try to find time for it. I cannot make any promises regarding the timeframe, though.

Just to pick up one of your original questions: Yes, you can share a socket between containers if they are using the sidecar pattern. I assume you're doing that inside Kubernetes? Then you can define the folder where the socket should be placed as a volume and mount that in both containers within the same pod 😄

**Comment by claytondaley** [comment]
Thanks.  I had tried (and failed miserably) to get the TCP socket working with a sidecar.  When I finally gave up, I decided to echo the aspiration here.

When I circled back, I figured out that my issue was just `localhost` vs `0.0.0.0` so it is now working on our end.  A properly configured TCP socket doesn't require the shared volume which is convenient.  Obviously, what we learned won't help you if you don't go the supervisor route, but may help someone else in the meantime.
